thewalrus/internal_modes/pnr_statistics.py

```diff
@@ -0,0 +1,122 @@
+# Copyright 2021-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Set of functions for calculating photon number resolved measurement probabilities on Gaussian states with multiple internal modes
+"""
+
+import numpy as np
+
+import numba
+
+from scipy.special import factorial as fac
+
+from ..quantum import Qmat
+from .._hafnian import find_kept_edges, nb_binom, f_from_powertrace, nb_ix
+from ..charpoly import powertrace
+
+from .useful_tools import spatial_reps_to_schmidt_reps, spatial_modes_to_schmidt_modes
+
+
+@numba.jit(nopython=True, parallel=True, cache=True)
+def hafkd(As, edge_reps, K=1):
+    r"""
+    generalised version of hafnian to include multiple internal modes
+
+    Args:
+        As (array[:,:,:]): :math:`D\times2MK\times2MK` array of D covariance matrices, each being a 2MK covariance matrix
+        edge_reps (array[int]): length-:math:M array. Gives how many times edges are repeated, or how many photons in each mode
+        K (int): How many internal modes are in each detected mode
+    """
+
+    M = As[0].shape[0] // (2 * K)
+
+    N = 2 * edge_reps.sum()
+
+    steps = np.prod(edge_reps + 1)
+
+    H = 0.0
+
+    for j in numba.prange(steps):
+        kept_edges = find_kept_edges(j, edge_reps)
+        edges_sum = kept_edges.sum()
+
+        binom_prod = 1.0
+        for i in range(M):
+            binom_prod *= nb_binom(edge_reps[i], kept_edges[i])
```

Comment:
```suggestion
            binom_prod *= nb_binom(np.int64(edge_reps[i]), np.int64(kept_edges[i]))
```
Again for numba to work I need to change this or inside `nb_binom`.